### DISCUSSION TITLE
State and Async in Error cases

### DISCUSSION THREAD

**Discussion by ibotty** [tokio-rs/axum#2902] [Q&A]
### Summary

Hi,

I'd like to template errors.  Transforming the errors in `IntoResponse` works perfectly, but I'd like to have some state (URL, etc) in the response and would like to use async libraries.

The gist of the app ATM is.

```rust
struct AppError {};

impl IntoResponse for AppError {...}

async fn get_root() -> Result<(), AppError> {
  Err(AppError)
}
 
let app = axum::Router::new().route("/", get(get_root));
axum::serve(TcpListener::bind("0.0.0.0:8080"), app);
```

I tried to use `HandleError` and `HandleErrorLayer`, but could not get it to compose with the usual routing in axum.

Can someone give a a hint?

### axum version

0.7.4

**Comment by randomairborne** [comment]
Usually, HandleErrorLayer is for something different. When I work with errors that need to be templated, I do something a little weird.
I have my IntoResponse implementation insert an extension with my error type into the http [extension](https://docs.rs/http/latest/http/response/struct.Response.html#method.extensions) field, and in a middleware, I check if that extension exists, and form a separate "error response" there if so, discarding the original response.

[Example IntoResponse impl](https://github.com/randomairborne/mcping/blob/c78873bfaf9e0817c159905f9044cddf6bb1a023/src/main.rs#L435-L449)
[Example middleware](https://github.com/randomairborne/mcping/blob/c78873bfaf9e0817c159905f9044cddf6bb1a023/src/main.rs#L514-L551)
